### ISSUE TITLE
Add support for slim jars

### ISSUE THREAD

**Issue by aslo** [GoogleCloudPlatform/runtime-builder-java#14]
Slim jars reference other local dependencies - these should be packaged in the runtime as well.

**Comment by aslo** [comment]
Actually gcloud currently uploads the entire working directory where "gcloud app deploy" is run, and that directory is what the builder operates on.

From the builder's perspective, our code would be responsible for examining a jar's manifest and locating the dependency jars within the directory.

From the tooling perspective, any tool that does staging would just need to stage any dependency jars within the deployed directory, and make sure that the main jar's manifest correctly points to them.

**Comment by chanseokoh** [comment]
But as I said in the email thread, currently if there are multiple JARs in a staging directory, it fails. I know it does upload the entire working directory if it is an exploded WAR, but I don't think it does when there are multiple JARs only.

**Comment by chanseokoh** [comment]
Ah, you are saying the current error output is by the builder and not ny gcloud?

**Comment by aslo** [comment]
Yes, exactly. The builder is what handles any language-specific processing and packaging of the user's code, and that's where the error comes from.

So, to make this work, the builder just needs to recognize slim-jar-with-dependencies deployments.

**Comment by chanseokoh** [comment]
I think the IntelliJ team will want this badly, because if this is supported, they don't have to do any extra processing at all. Eclipse can also benefit greatly, but we may not wait until this is ready but opt for building a fat JAR. @patflynn @etanshaul

**Comment by patflynn** [comment]
@aslo another option would be to only support fat jars.

**Comment by aslo** [comment]
@patflynn Well, that's effectively what we have now, since this feature hasn't been implemented.

I think this feature makes sense, and seems to be something that devrel is excited about.

**Comment by patflynn** [comment]
@aslo I'm just wondering if users really care. Is building a fat jar onerous for them? Are there use-cases that fat jars don't support maybe?

**Comment by aslo** [comment]
cc @lesv, who has been a big proponent of this - he's probably better suited to answer that.

One reason customers might care - fat jars can get quite heavy. Typically, application code changes a lot more frequently than the dependencies. If dependencies and user code are separated, they can be added to the docker image in separate layers, which allows docker's caching to minimize the bits it needs to send over the network.

See also http://product.hubspot.com/blog/the-fault-in-our-jars-why-we-stopped-building-fat-jars

**Comment by patflynn** [comment]
That's a fair point.  I'm guessing this is for non web-serving use-cases,
and it sounds generally useful for potential k8s Java workloads.

**Comment by lesv** [comment]
@patflynn Even SpringBoot has [issues](http://bsideup.blogspot.com/2015/04/spring-boots-fat-jars-vs-docker.html).  Ray was the one who suggested SlimJar's, I came to agree w/ him.
